### PR TITLE
add person sensor (SEN21231) from usefulsensors

### LINKED ISSUE #1925: Add support for the Person Sensor by Useful Sensors (SEN-21231)
**Describe the problem you have/What new integration you would like**
Please add support for the Person Sensor by Useful Sensors (SEN-21231) so that we can have good presence detention using ML sensor

**Please describe your use case for this integration and alternatives you've tried:**
 i've been using IR and Microwave sensor for presence detection but it is not that reliable. using this sensor with built in ML can elevate presence sensor to the next level. And the cost of the sensor is relatively affordable, USD 10 

**Additional context**
 module info : https://github.com/usefulsensors/person_sensor_docs

### PR DESCRIPTION
**Pull request by shreyaskarnik**
# What does this implement/fix?

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes https://github.com/esphome/feature-requests/issues/1925

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2678

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
i2c:
  sda: 21
  scl: 22
sensor:
  - platform: sen21231
  - id: person_sensor
    name: "Person Sensor"

## can also turn this into a binary sensor as a motion detection sensor 
binary_sensor:
  - platform: status
    name: "Sense Status"
  - platform: template
    name: Person Sensor Binary
    device_class: motion
    lambda: |-
      if (id(person_sensor).state) {
        return true;
      } else {
        return false;
      }

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs). https://github.com/esphome/esphome-docs/pull/2678

### CODE REVIEW

File: esphome/components/sen21231/sen21231.cpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
+#include "sen21231.h"
+#include "esphome/core/log.h"
+
+namespace esphome {
+namespace sen21231_sensor {
+
+static const char *TAG = "sen21231_sensor.sensor";
```
Comment (jesserockz) [inline esphome/components/sen21231/sen21231.cpp]:
```suggestion
static const char *const TAG = "sen21231_sensor.sensor";
```

File: esphome/components/sen21231/sen21231.cpp [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+void Sen21231Sensor::setup() {}
+
```
Comment (jesserockz) [inline esphome/components/sen21231/sen21231.cpp]: No need for an empty function
```suggestion
```

File: esphome/components/sen21231/sen21231.cpp [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+void Sen21231Sensor::update() { this->read_data_(); }
+
+void Sen21231Sensor::dump_config() {
+  ESP_LOGCONFIG(TAG, "SEN21231:");
+  LOG_I2C_DEVICE(this);
+  if (this->is_failed()) {
+    ESP_LOGE(TAG, "Communication with SEN21231 failed!");
+  }
+  ESP_LOGI(TAG, "SEN21231: %s", this->is_failed() ? "FAILED" : "OK");
+  this->read_data_();
```
Comment (jesserockz) [inline esphome/components/sen21231/sen21231.cpp]: Dont read during the dump please
```suggestion
```

File: esphome/components/sen21231/sen21231.cpp [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+}
```
Comment (jesserockz) [inline esphome/components/sen21231/sen21231.cpp:19]:
```suggestion
 LOG_UPDATE_INTERVAL(this);
}
```

File: esphome/components/sen21231/sen21231.cpp [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+void Sen21231Sensor::read_data_() {
+  person_sensor_results_t results;
+  this->read_bytes(PERSON_SENSOR_I2C_ADDRESS, (uint8_t *) &results, sizeof(results));
+  ESP_LOGD(TAG, "SEN21231: %d faces detected", results.num_faces);
+  if (results.num_faces > 0) {
+    this->publish_state(true);
+  } else {
+    this->publish_state(false);
+  }
```
Comment (jesserockz) [inline esphome/components/sen21231/sen21231.cpp]: This is a sensor, it should publish a number...
```suggestion
  this->publish_state(results.num_faces);
```

File: esphome/components/sen21231/sen21231.h [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+// verify this in practice, but we found it useful during our own debugging.
+typedef struct __attribute__((__packed__)) {
+  person_sensor_results_header_t header;                      // Bytes 0-4.
+  int8_t num_faces;                                           // Byte 5.
+  person_sensor_face_t faces[PERSON_SENSOR_MAX_FACES_COUNT];  // Bytes 6-37.
+  uint16_t checksum;                                          // Bytes 38-39.
+} person_sensor_results_t;
+class Sen21231Sensor : public sensor::Sensor, public PollingComponent, public i2c::I2CDevice {
+ public:
+  void setup() override;
```
Comment (jesserockz) [inline esphome/components/sen21231/sen21231.h]: Not needed
```suggestion
```

File: esphome/components/sen21231/sen21231.h [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+  void update() override;
+  void dump_config() override;
+  void read_data_();
```
Comment (jesserockz) [inline esphome/components/sen21231/sen21231.h:73]:
```suggestion
 protected:
  void read_data_();
```

File: esphome/components/sen21231/sen21231.h [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+// sensor when we do an I2C read from the peripheral address.
+// The checksum should be the CRC16 of bytes 0 to 38. You shouldn't need to
+// verify this in practice, but we found it useful during our own debugging.
+typedef struct __attribute__((__packed__)) {
+  person_sensor_results_header_t header;                      // Bytes 0-4.
+  int8_t num_faces;                                           // Byte 5.
+  person_sensor_face_t faces[PERSON_SENSOR_MAX_FACES_COUNT];  // Bytes 6-37.
+  uint16_t checksum;                                          // Bytes 38-39.
+} person_sensor_results_t;
+class Sen21231Sensor : public sensor::Sensor, public PollingComponent, public i2c::I2CDevice {
```
Comment (jesserockz) [inline esphome/components/sen21231/sen21231.h:67]: Readability
```suggestion

class Sen21231Sensor : public sensor::Sensor, public PollingComponent, public i2c::I2CDevice {
```

File: esphome/components/sen21231/sensor.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    sensor.sensor_schema(
+        unit_of_measurement=UNIT_EMPTY, icon=ICON_MOTION_SENSOR, accuracy_decimals=1
+    )
+    .extend(
+        {
+            cv.GenerateID(): cv.declare_id(Sen21231Sensor),
+        }
+    )
+    .extend(cv.polling_component_schema("30s"))
+    .extend(i2c.i2c_device_schema(CONF_I2C_ADDR))
```
Comment (jesserockz) [inline esphome/components/sen21231/sensor.py]: No need for a new CONST for the address
```suggestion
    .extend(i2c.i2c_device_schema(0x62))
```

File: esphome/components/sen21231/sensor.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
+import esphome.codegen as cg
+import esphome.config_validation as cv
+from esphome.components import i2c, sensor
+from esphome.const import CONF_ID, ICON_MOTION_SENSOR, UNIT_EMPTY
+
+DEPENDENCIES = ["i2c"]
+
+CONF_I2C_ADDR = 0x62
+
```
Comment (jesserockz) [inline esphome/components/sen21231/sensor.py]:
```suggestion
```

File: esphome/components/sen21231/sensor.py [hunk 11]
```diff
@@ -0,0 +1,33 @@
+import esphome.codegen as cg
+import esphome.config_validation as cv
+from esphome.components import i2c, sensor
+from esphome.const import CONF_ID, ICON_MOTION_SENSOR, UNIT_EMPTY
+
+DEPENDENCIES = ["i2c"]
```
Comment (jesserockz) [inline esphome/components/sen21231/sensor.py:7]:
```suggestion
CODEOWNERS = ["@shreyaskarnik"]
DEPENDENCIES = ["i2c"]
```

File: esphome/components/sen21231/sensor.py [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    .extend(cv.polling_component_schema("30s"))
+    .extend(i2c.i2c_device_schema(CONF_I2C_ADDR))
+)
+
+
+def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    yield cg.register_component(var, config)
+    yield sensor.register_sensor(var, config)
+    yield i2c.register_i2c_device(var, config)
```
Comment (jesserockz) [inline esphome/components/sen21231/sensor.py]:
```suggestion
async def to_code(config):
    var = await sensor.new_sensor(config)
    await cg.register_component(var, config)
    await i2c.register_i2c_device(var, config)
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

**Comment by shreyaskarnik** [comment]
Thank you for your feedback @jesserockz really really appreciate it. Changes requested have been applied and tested on ESP32 based system.

**Comment by SeByDocKy** [comment]
Sound a great addition :). How good is this sensor ?

**Comment by shreyaskarnik** [comment]
@SeByDocKy the sensor works pretty well in good lighting conditions. So far I've had no issues.

**Comment by SeByDocKy** [comment]
> @SeByDocKy the sensor works pretty well in good lighting conditions. So far I've had no issues.

Ok I ordered two of them.... to try your PR :)

File: esphome/components/sen21231/sensor.py [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+CONFIG_SCHEMA = (
+    sensor.sensor_schema(
+        unit_of_measurement=UNIT_EMPTY, icon=ICON_MOTION_SENSOR, accuracy_decimals=1
+    )
+    .extend(
+        {
+            cv.GenerateID(): cv.declare_id(Sen21231Sensor),
+        }
+    )
```
Comment (jesserockz) [inline esphome/components/sen21231/sensor.py]:
```suggestion
    sensor.sensor_schema(
        Sen21231Sensor, icon=ICON_MOTION_SENSOR, accuracy_decimals=1
    )
```

File: esphome/components/sen21231/sensor.py [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+    .extend(cv.polling_component_schema("30s"))
```
Comment (jesserockz) [inline esphome/components/sen21231/sensor.py]: We always try to default sensors to 60s unless there is a specific reason? Its easy for a user to override this in yaml.
```suggestion
    .extend(cv.polling_component_schema("60s"))
```

File: esphome/components/sen21231/sen21231.h [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+typedef struct __attribute__((__packed__)) {
+  uint8_t box_confidence;  // Byte 1.
+  uint8_t box_left;        // Byte 2.
+  uint8_t box_top;         // Byte 3.
+  uint8_t box_right;       // Byte 4.
+  uint8_t box_bottom;      // Byte 5.
+  int8_t id_confidence;    // Byte 6.
+  int8_t id;               // Byte 7
+  uint8_t is_facing;       // Byte 8.
+} person_sensor_face_t;
```
Comment (jesserockz) [inline esphome/components/sen21231/sen21231.h]:
```suggestion
using person_sensor_face_t = struct __attribute__((__packed__)) {
  uint8_t box_confidence;  // Byte 1.
  uint8_t box_left;        // Byte 2.
  uint8_t box_top;         // Byte 3.
  uint8_t box_right;       // Byte 4.
  uint8_t box_bottom;      // Byte 5.
  int8_t id_confidence;    // Byte 6.
  int8_t id;               // Byte 7
  uint8_t is_facing;       // Byte 8.
};
```

File: esphome/components/sen21231/sen21231.h [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+// This is the full structure of the packet returned over the wire from the
+// sensor when we do an I2C read from the peripheral address.
+// The checksum should be the CRC16 of bytes 0 to 38. You shouldn't need to
+// verify this in practice, but we found it useful during our own debugging.
+typedef struct __attribute__((__packed__)) {
+  person_sensor_results_header_t header;                      // Bytes 0-4.
+  int8_t num_faces;                                           // Byte 5.
+  person_sensor_face_t faces[PERSON_SENSOR_MAX_FACES_COUNT];  // Bytes 6-37.
+  uint16_t checksum;                                          // Bytes 38-39.
+} person_sensor_results_t;
```
Comment (jesserockz) [inline esphome/components/sen21231/sen21231.h]:
```suggestion
using person_sensor_results_t = struct __attribute__((__packed__)) {
  person_sensor_results_header_t header;                      // Bytes 0-4.
  int8_t num_faces;                                           // Byte 5.
  person_sensor_face_t faces[PERSON_SENSOR_MAX_FACES_COUNT];  // Bytes 6-37.
  uint16_t checksum;                                          // Bytes 38-39.
};
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

**Comment by shreyaskarnik** [comment]
Thanks @jesserockz for the suggestions I have applied all of them. Really appreciate your feedback and time to review the PR! 🙌

**Review by jesserockz — APPROVED** [review]

**Pull request merged**